### PR TITLE
feat: Add compaction improvements and manual controls

### PR DESCRIPTION
**Pull request by clusterfudge**
## Overview

This PR introduces several enhancements to conversation compaction, addressing issues where compaction was triggering too late and providing users with better control and visibility.

### 1. Lower Default Compaction Threshold (85% → 65%)
- **Before**: Triggered at ~170K tokens (440 messages)
- **After**: Triggers at ~130K tokens (340 messages)
- More reasonable trigger point for practical conversations
- Prevents conversations from growing too large before compaction

### 2. Environment Variable Configuration
New environment variables for flexible configuration:
- `SILICA_COMPACTION_THRESHOLD`: Configure threshold without code changes (e.g., 0.50 for 50%)
- `SILICA_DEBUG_COMPACTION`: Enable detailed compaction debugging output

### 3. Manual Compaction CLI Commands
Two new commands for explicit user control:
- `/compact`: Full conversation compaction with archiving
- `/mc [N]`: Micro-compact first N turns (default 3) while keeping the rest

**Note on turn calculation**: A turn must end with a user message per API requirements:
- Turn 1 = 1 message (user)
- Turn 2 = 3 messages (user, assistant, user)
- Turn 3 = 5 messages (user, assistant, user, assistant, user)
- Turn N = (2N - 1) messages

### 4. Enhanced Debugging & Error Handling
- Detailed token usage statistics in debug mode
- Clear compaction decision logging
- Better error messages with full stack traces
- Visibility into why compaction triggers or doesn't trigger

### 5. Path Handling Fix
- Fixed `_get_history_dir()` to handle both string and Path objects properly

## Files Changed

- **silica/developer/compacter.py**: Threshold, env vars, debug mode
- **silica/developer/agent_loop.py**: Debug integration
- **silica/developer/context.py**: Path handling fix
- **silica/developer/toolbox.py**: Manual compaction commands with correct turn calculation
- **.env.example**: Document new environment variables
- **tests/developer/test_compaction_commands.py**: Test new commands
- **docs/developer/compaction_commands.md**: Command reference with correct turn calculations
- **docs/developer/compaction_improvements.md**: Comprehensive documentation

## Commits

1. **cd0ba5c**: Initial compaction improvements
2. **2545139**: Fix turn calculation in micro-compact (was incorrectly N×2, now correctly (2N-1))

## Testing

All 29 compaction tests pass:
```bash
pytest tests/developer/test_compaction*.py -v
======================== 29 passed in 0.45s =========================
```

Test coverage includes:
- Core compaction functionality
- Timing and trigger conditions
- Session ID preservation
- Message validation
- New manual compaction commands with correct turn calculations
- Threshold behavior

### Default Behavior (Recommended)
No configuration needed. Compaction triggers at 65%:
```bash
silica
```

### Custom Threshold
```bash
# More aggressive (50%)
export SILICA_COMPACTION_THRESHOLD=0.50
silica

# More conservative (80%)
export SILICA_COMPACTION_THRESHOLD=0.80
silica
```

### Debugging
```bash
export SILICA_DEBUG_COMPACTION=1
silica
```

Output:
```
[Compaction Check]
  Model: claude-3-5-sonnet-20241022
  Context window: 200,000
  Threshold ratio: 65%
  Token threshold: 130,000
  Current tokens: 38,350
  Usage: 19.2%
  Should compact: False
```

### Manual Compaction
```bash
# Full compaction
> /compact
✓ Conversation compacted successfully!
**Original:** 240 messages (85,000 tokens)
**Compacted:** 3 messages (1,500 tokens)
**Archive:** pre-compaction-20251017_143022.json

# Micro-compaction (default 3 turns = 5 messages)
> /mc
✓ Micro-compaction completed!
**Compacted:** First 3 turns (5 messages)
**Kept:** 9 messages from the rest
**Final message count:** 10 (was 14)

# Micro-compaction (custom 5 turns = 9 messages)
> /mc 5
✓ Micro-compaction completed!
**Compacted:** First 5 turns (9 messages)
**Kept:** 21 messages from the rest
**Final message count:** 22 (was 30)
```

### For Users
- **Better Control**: Manual compaction commands for strategic compression
- **Flexibility**: Environment variables for easy threshold adjustment
- **Transparency**: Debug mode shows exactly what's happening
- **Optimization**: Proactive token usage management
- **Correct Behavior**: Turn calculations now match API requirements

### For Developers
- **Reusability**: Leverages existing compaction infrastructure
- **Testing**: Manual triggers make testing easier
- **Debugging**: Clear visibility into compaction behavior
- **Maintenance**: Minimal new code, follows existing patterns
- **Correctness**: Proper turn counting for API compliance

### Breaking Changes
None. All changes are backward compatible.

### Behavior Changes
- Compaction triggers earlier (65% vs 85%)
- Generally an improvement, but configurable via environment variable
- Micro-compact now correctly calculates turns as (2N-1) messages

### Reverting to Old Behavior
```bash
export SILICA_COMPACTION_THRESHOLD=0.85
```

## Related Work
- Session ID preservation (#56)
- Keyboard interrupt handling (#50)
- Extended thinking API (#37)

## Documentation
- Full command reference: `docs/developer/compaction_commands.md`
- Implementation details: `COMPACTION_FIX_SUMMARY.md`
- Usage guide: `COMPACTION_COMMANDS_SUMMARY.md`

### CODE REVIEW

File: silica/developer/agent_loop.py [hunk 1]
```diff
@@ -246,16 +246,35 @@ def _check_and_apply_compaction(
...
         return agent_context, False
 
     # Only check compaction when conversation state is complete
     # (no pending tool results and conversation has actual content)
-    if (
-        agent_context.tool_result_buffer
-        or not agent_context.chat_history
-        or len(agent_context.chat_history) <= 2
-    ):
+    if agent_context.tool_result_buffer:
```
Comment (clusterfudge) [inline silica/developer/agent_loop.py]: We need to fix this, if we got into a loop of tool uses, we might overshoot the compaction threshold and crash. we should convert the tool result buffer into messages first, and then check if we should compact.

File: silica/developer/toolbox.py [hunk 2]
```diff
@@ -1060,6 +1082,157 @@ def _model(self, user_interface, sandbox, user_input, *args, **kwargs):
...
+                markdown=False,
+            )
+
+            # Convert messages to string for summarization
+            conversation_str = compacter._messages_to_string(
+                messages_to_summarize, for_summary=True
+            )
+
+            # Generate summary using Claude
+            system_prompt = f"""
```
Comment (clusterfudge) [inline silica/developer/toolbox.py]: no, don't reimplement compaction and the prompt. reuse the existing compactor implementation and simply pass it the subset of the conversation.

**Comment by clusterfudge** [comment]
## Review Feedback Addressed

All review comments have been addressed in commit a325d50:

### ✅ Documentation Cleanup
- Removed `COMPACTION_FIX_SUMMARY.md`
- Removed `COMPACTION_COMMANDS_SUMMARY.md`
- Removed `docs/developer/compaction_improvements.md`
- Only `docs/developer/compaction_commands.md` remains (sufficient documentation)

### ✅ Threshold Restored to 0.80
Changed from 0.65 back to 0.80 in `DEFAULT_COMPACTION_THRESHOLD_RATIO`

### ✅ Variable Rename
Fixed typo: `preserver_message_count` → `preserve_message_count`

### ✅ CRITICAL: Compaction Check Timing Fix
**This was the most important change:**
- Moved compaction check in `agent_loop.py` from line 408 (before tool conversion) to after line 495 (after tool_result_buffer is converted to messages)
- Prevents overshooting threshold during loops of tool uses
- Ensures complete conversation state including all tool interactions before checking
- Eliminates risk of crashing due to exceeding context window mid-tool-loop

### ✅ Reuse Existing Compactor in /mc Command
- No longer reimplements summarization and prompt logic
- Creates temporary `AgentContext` with messages to compact
- Calls existing `generate_summary()` method
- Properly uses `CompactionSummary` object structure

### ✅ Test Fix
Updated edge case test expectations for 2-message compaction scenario

---

**All 29 compaction tests pass** ✅

**Comment by clusterfudge** [comment]
## ✅ CI Fixed!

**Commit 4a04c40** fixes the final CI test failures.

### Problem
Tests for `/compact` and `/mc` commands were failing in CI because:
- These commands create a new Anthropic client in toolbox.py
- They call `os.getenv("ANTHROPIC_API_KEY")` which returns None in CI
- This caused "ANTHROPIC_API_KEY environment variable not set" errors

### Solution
Added `@mock.patch.dict("os.environ", {"ANTHROPIC_API_KEY": "test-key"})` to the three affected tests:
- test_compact_command
- test_micro_compact_command_default
- test_micro_compact_command_custom_turns

### Result
✅ **All CI checks now pass!**
✅ All 29 compaction tests pass without real API key
✅ PR is ready for merge
